src/lib/pubkey/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,816 @@
+      constexpr void conditional_assign(bool cond, const Self& other) {
+         CT::conditional_assign_mem(static_cast<W>(cond), m_val.data(), other.data(), N);
+      }
+
+      // TODO be faster
+      constexpr Self dbl() const { return (*this) + (*this); }
+
+      constexpr Self square() const {
+         std::array<W, 2 * N> z;
+         comba_sqr<N>(z.data(), this->data());
+         return bigint_monty_redc(z, Self::P, Self::P_dash);
+      }
+
+      // Negation modulo p
+      constexpr Self negate() const {
+         auto x_is_zero = CT::all_zeros(this->data(), N);
+
+         std::array<W, N> r;
+         bigint_sub3(r.data(), Self::P.data(), N, this->data(), N);
+         x_is_zero.if_set_zero_out(r.data(), N);
+         return Self(r);
+      }
+
+      constexpr Self pow_vartime(const std::array<W, N>& exp) const {
+         auto x = (*this);
+         auto y = Self::one();
+
+         for(size_t i = 0; i != Self::BITS; ++i) {
+            if(get_bit(i, exp)) {
+               y = y * x;
+            }
+            x = x.square();
+         }
+
+         return y;
+      }
+
+      /**
+      * Returns the modular inverse, or 0 if no modular inverse exists.
+      *
+      * If the modulus is prime the only value that has no modular inverse is 0.
+      *
+      * This uses Fermat's little theorem, and so assumes that p is prime
+      */
+      constexpr Self invert() const { return pow_vartime(Self::P_MINUS_2); }
+
+      /**
+      * Return the modular square root, or zero if no root exists
+      *
+      * Current impl assumes p == 3 (mod 4)
+      */
+      constexpr Self sqrt() const {
+         static_assert(Self::P_MOD_4 == 3);
+         auto z = pow_vartime(Self::P_PLUS_1_OVER_4);
+         const bool correct = (z * z) == *this;
+         z.conditional_assign(!correct, Self::zero());
+         return z;
+      }
+
+      constexpr bool is_square() const {
+         auto z = pow_vartime(Self::P_MINUS_1_OVER_2);
+         const bool is_one = z.is_one();
+         const bool is_zero = z.is_zero();
+         return (is_one || is_zero);
+      }
+
+      constexpr bool operator==(const Self& other) const {
+         return CT::is_equal(this->data(), other.data(), N).as_bool();
+      }
+
+      constexpr bool operator!=(const Self& other) const {
+         return CT::is_not_equal(this->data(), other.data(), N).as_bool();
+      }
+
+      constexpr std::array<uint8_t, Self::BYTES> serialize() const {
+         auto v = bigint_monty_redc(m_val, Self::P, Self::P_dash);
+         std::reverse(v.begin(), v.end());
+         auto bytes = store_be(v);
+
+         if constexpr(Self::BYTES == Self::N * WordInfo<W>::bytes) {
+            return bytes;
+         } else {
+            // Remove leading zero bytes
+            const size_t extra = Self::N * WordInfo<W>::bytes - Self::BYTES;
+            std::array<uint8_t, Self::BYTES> out;
+            copy_mem(out.data(), &bytes[extra], Self::BYTES);
+            return out;
```

Comment:
Perhaps using range-based `copy_mem`?

```suggestion
            // Remove leading zero bytes
            constexpr size_t extra = Self::N * WordInfo<W>::bytes - Self::BYTES;
            std::array<uint8_t, Self::BYTES> out;
            copy_mem(out, std::span{bytes}.template subspan<extra, Self::BYTES>();
            return out;
```

Reply:
:+1: